### PR TITLE
Allow enabling standby mode on instances without EIP's

### PR DESCRIPTION
**Pull request by sevenmachines**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -103,4 +103,26 @@ a more resilient and automated solution than assigning EIP's manually.
...
+
+Note if the standby mode function is enabled, this requires an additional set of IAM permissions.
+The following EC2 permissions are required.
+
+.. code::
+
+	"ec2:AllocateAddress",
+  "ec2:AssociateAddress",
+  "ec2:DescribeAddresses",
+  "ec2:DisassociateAddress"
```
Comment (lgarvey) [inline README.rst]: Does the script ever need to dissociate the EIP?

File: README.rst [hunk 2]
```diff
@@ -103,4 +103,26 @@ a more resilient and automated solution than assigning EIP's manually.
...
+    # Enable or disable the use of standby mode to add and remove 
+    # instances automatically dependent on whether thay have 
+    # acquired an EIP or not
+    eip_enable_standby_mode: True
+
+
+Note if the standby mode function is enabled, this requires an additional set of IAM permissions.
+The following EC2 permissions are required.
+
+.. code::
```
Comment (lgarvey) [inline README.rst:115]: I'm wondering if it's worth actually including the policy to ease set-up, or perhaps add the file to the formula repo with some instructions on including it via b-cfn's include directive.

``` json
{
    "Version": "2012-10-17", 
    "Statement": [ 
        { 
            "Action": [ 
                "ec2:AllocateAddress", 
                "ec2:AssociateAddress",
                "ec2:DescribeAddresses",
                "autoscaling:EnterStandby",
                "autoscaling:ExitStandby"
            ],
            "Resource": [
                "*"
            ],
            "Effect": "Allow"
        }
    ]
}
```

File: README.rst (hunk 1 above)
Comment (sevenmachines) [reply README.rst]: the standy mode in ASG associates and disassociates from the ELB in the background

File: README.rst (hunk 2 above)
Comment (sevenmachines) [reply README.rst:115]: Sounds good, I was going to fit it in as an iam-autoeip.json file in template-deploy for optional inclusion, but not sure if thats the best way?

File: README.rst (hunk 1 above)
Comment (lgarvey) [reply README.rst]: Admittedly I haven't done any tests to verify that the standby functionality works [will do that in a bit!] but wouldn't the perms required be: 

elasticloadbalancing:RegisterInstancesFromLoadBalancer
elasticloadbalancing:DeregisterInstancesFromLoadBalancer

File: README.rst (hunk 2 above)
Comment (lgarvey) [reply README.rst:115]: That sounds great.

File: README.rst (hunk 1 above)
Comment (sevenmachines) [reply README.rst]: oh, sorry. totally misunderstood. Yep, you could probably drop the dissocate permission, unless we can think of a probable useful use case? Allocate address isnt required either

Comment (sevenmachines) [reply README.rst]: so yes, AssociateAddress and describe only I think?

**Comment by lgarvey** [comment]
Yep, agreed.

File: README.rst (hunk 1 above)
Comment (sevenmachines) [reply README.rst]: Updated the PR

File: aws/autoeips.sls [hunk 3]
```diff
@@ -8,8 +8,11 @@ autoeips.py:
     - mode: 755
     - template: jinja
 
+boto3:
+  pkg.installed
```
Comment (lgarvey) [inline aws/autoeips.sls:12]: I suspect this should be: 

``` yaml
boto3:
  pip.installed:
    - user: root
```

File: aws/files/autoeips.py [hunk 4]
```diff
@@ -74,10 +120,21 @@ def associate_eip(self, eips, retries=3):
...
                 success = eip.associate(
                     instance_id=self.instance_id, allow_reassociation=False)
+                self.update_standby_mode(False)
                 return success
+        logger.warning("Failed to associate with any EIP's")
+        self.update_standby_mode(True)
         return False
 
     def get_unassociated_eips(self):
+        """
```
Comment (lgarvey) [inline aws/files/autoeips.py:130]: I've also noticed that, when testing, if I set filter_addresses to an empty list, then the code will grab any unassociated EIP. I think this is a bit dangerous if it happens by default. 

We probably just need something like: 

``` python
if not self.filter_address:
     logger.critical("EIPs must be explicitly set in the pillar.")
     self.safe_exit(1)
```

Comment (lgarvey) [reply aws/files/autoeips.py:130]: Just realised you already have some code to do that:

if (self.filter_addresses < 1):

The pythonic way is:

if not self.filter_addresses:

Or you could do it your way if you get the length of self.filter_addresses, e.g. len(self.filter_addresses)

**Comment by lgarvey** [comment]
I've tested the standby code and it's working fine.  I think it's just the minor issue with not checking len on filtered_addresses and the boto3 sls, and I'd be happy to merge.

File: aws/files/autoeips.py (hunk 4 above)
Comment (sevenmachines) [reply aws/files/autoeips.py:130]: yep, just noticed that tiredness failure :) If I can ignore the pythonic though please. I find len(self.filter_addresses) < 1 just _so_ much clearer.

File: aws/autoeips.sls (hunk 3 above)
Comment (sevenmachines) [reply aws/autoeips.sls:12]: I'll take your word for this one.

**Comment by sevenmachines** [comment]
@LyndonGarvey updated the PR with the 2 outstanding things

File: aws/files/autoeips.py (hunk 4 above)
Comment (lgarvey) [reply aws/files/autoeips.py:130]: I've got no issues with the more traditional syntax!
